### PR TITLE
feat: develop the training scripts to achieve compliance in velocity controlled robots

### PR DESCRIPTION
**Pull request by yrh012**
## Description

- #44 

This PR solves the issue by introducing an new gym environment that allows the training of a RL policy that aims to reach a certain target pose in space while avoiding exceeding a certain force limit. In this regard, a UR5e configuration has been develop that allow learning such policy using a UR5e robot. It also allows introducing a new type of clipping in task space that clips cartesian commands between an upper and lower limit. It also moves away from the USD file of the UR5e that is stored on the cloud to an locally saved one, allowing much faster training initiation.

## Review guidelines

Estimated Time of Review: 10 minutes

## Checklist before merging:

- [ ] Confirm that the relevant changelog(s) are up-to-date in case of any user-facing changes

### CODE REVIEW

**Review by bpapaspyros — COMMENTED** [review]
looks overall good, some general comments on my end. I do wonder about some reward weights (seem a bit imbalanced to me), but you have already tuned quite a bit so ultimately if it works, it works

File: source/isaaclab_tasks/isaaclab_tasks/manager_based/manipulation/control/config/ur5e/force_limit_cfg.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        self.commands.ee_pose.ranges.pitch = (0, 0)
+
+        # override actions
+        self.actions.arm_action = DifferentialInverseKinematicsActionCfg(
+            asset_name="robot",
+            joint_names=[".*"],
+            body_name=self.ee_str,
+            controller=DifferentialIKControllerCfg(command_type="velocity", ik_method="dls"),
+            scale=0.02,
+            clip=(-0.06, 0.06),
```
[the remark this answers is not in the document]
Comment (yrh012) [reply source/isaaclab_tasks/isaaclab_tasks/manager_based/manipulation/control/config/ur5e/force_limit_cfg.py:40]: Yep let me explain it, the output of the network initially samples values from a gaussian distribution and with the default scale of 1, the policy samples values between -3 and 3 and centered at 0. Since the output of the policy is directly feed as setpoints to the IK velocity controller, the values it might get in initial phases can be extremely fast. While we know to each force compliance the robot should move slow, so with the above values, the policy samples values between -6 cm/s and 6 cm/s (which is still reasonable fast on the limits), we can ensure that the training is much more efficient in the early stages making it train better.

File: source/isaaclab_tasks/isaaclab_tasks/manager_based/manipulation/control/force_limit_scene_cfg.py [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+            rigid_props=sim_utils.RigidBodyPropertiesCfg(kinematic_enabled=True, max_depenetration_velocity=0.1),
+            activate_contact_sensors=True,
+        ),
+        init_state=RigidObjectCfg.InitialStateCfg(pos=(0.8, 0.0, 0.15), rot=(1.0, 0.0, 0.0, 0.0)),
+    )
+
+    contact_sensor = ContactSensorCfg(
+        prim_path="{ENV_REGEX_NS}/Robot/wrist_3_link",
+        update_period=0.0,
+        history_length=2,
```
[the remark this answers is not in the document]
Comment (yrh012) [reply source/isaaclab_tasks/isaaclab_tasks/manager_based/manipulation/control/force_limit_scene_cfg.py:83]: It always gives a matrix, a matrix of history_length by 3 so in this case, I'll make a matrix 2x3, (it is not filtered) and updated on physics tick as the update period is 0.0

**Comment by bpapaspyros** [comment]
> > looks overall good, some general comments on my end. I do wonder about some reward weights (seem a bit imbalanced to me), but you have already tuned quite a bit so ultimately if it works, it works
> 
> Good Observation! Yes, in a multi-objective RL, I ironically found out that the rewards weighting like this works best.

that's the point though, I don't think Isaac Sim is doing true multi-objective, but rather sums reward terms in a single weighted tensor (but I haven't gone deep to check). That's why I would imagine small gains are under-represented in the search space. But again, if it works for now that's ok, just a note for future tests where we may need more diverse solutions

**Comment by bpapaspyros** [comment]
> Yes by multi-objective I am really talking about the fact that I expect the model to satisfy multiple objectives and not the actual definition in optimization, in a sense that:
> 
> -1- I want the policy to guide a robot to a pose -2- I want the policy to avoid exceeding force limits while doing that -3- I want the policy to push in the direction of the pose error
> 
> If all have similar weights what would happen is that since it is not true multi-objective optimization, it will stay in a local minima that achieves the above "half-well" on all, but if we start valuing one over another, in a sense that I want the to be very accurate on pose tracking (less than 1cm error) then it would not to be valued more compared to other objectives and other objectives would play a role in finetuning that specific search space.
> 
> You have a really fair point that about the fact why small weights sometimes impact the policy too much and that's something worth investigating for sure!

sure, my second comment is not about the actual values, but on the definitions of multi-objective, we still optimize a single function so it's really single-objective with whatever that comes with. In the case of single objective, I would imagine you have to have these imbalanced weights at the risk of under-representing some things. 

The reason I am bringing it up is that if we start rewarding more functionality (directions, more randomization, etc) we may hit a wall with traditional single-objective

**Comment by yrh012** [comment]
> > Yes by multi-objective I am really talking about the fact that I expect the model to satisfy multiple objectives and not the actual definition in optimization, in a sense that:
> > -1- I want the policy to guide a robot to a pose -2- I want the policy to avoid exceeding force limits while doing that -3- I want the policy to push in the direction of the pose error
> > If all have similar weights what would happen is that since it is not true multi-objective optimization, it will stay in a local minima that achieves the above "half-well" on all, but if we start valuing one over another, in a sense that I want the to be very accurate on pose tracking (less than 1cm error) then it would not to be valued more compared to other objectives and other objectives would play a role in finetuning that specific search space.
> > You have a really fair point that about the fact why small weights sometimes impact the policy too much and that's something worth investigating for sure!
> 
> sure, my second comment is not about the actual values, but on the definitions of multi-objective, we still optimize a single function so it's really single-objective with whatever that comes with. In the case of single objective, I would imagine you have to have these imbalanced weights at the risk of under-representing some things.
> 
> The reason I am bringing it up is that if we start rewarding more functionality (directions, more randomization, etc) we may hit a wall with traditional single-objective

I completely agree with this point

File: source/isaaclab_tasks/isaaclab_tasks/manager_based/manipulation/control/mdp/rewards.py [hunk 3]
```diff
@@ -109,10 +109,110 @@ def force_limit_penalty(
...
+    asset_cfg: SceneEntityCfg,
+    contact_sensor_cfg: SceneEntityCfg = SceneEntityCfg("contact_sensor"),
+    command_name: str = "ee_pose",
+    limit: float = 1,
+) -> torch.Tensor:
+    """Reward the force in the direction of the position error."""
+    force = measured_forces_in_world_frame(env, contact_sensor_cfg)
+    position_error = position_command_error(env, command_name, asset_cfg, raw_vector=True)
+
+    force_position = torch.sum(force * position_error, dim=1)
```
Comment (eeberhard) [inline source/isaaclab_tasks/isaaclab_tasks/manager_based/manipulation/control/mdp/rewards.py:199]: In `force * position_error`, what is the product `*` of `torch.Tensor`s? There are many meanings to tensor multiplication (dot, cross, hadamard, outer product...)

You are doing a sum implying the result is not a scalar. I thought we discussed it as the dot product.

In any case, it would be more readable using the corresponding named method `dot`, `mul` etc

File: source/isaaclab_tasks/isaaclab_tasks/manager_based/manipulation/control/force_limit_scene_cfg.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (bpapaspyros) [reply source/isaaclab_tasks/isaaclab_tasks/manager_based/manipulation/control/force_limit_scene_cfg.py:83]: ok, that makes more sense now. And is just 2 samples enough? Do we get anything sensible if we go with more?

File: source/isaaclab_tasks/isaaclab_tasks/manager_based/manipulation/control/config/ur5e/force_limit_cfg.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (bpapaspyros) [reply source/isaaclab_tasks/isaaclab_tasks/manager_based/manipulation/control/config/ur5e/force_limit_cfg.py:40]: ok, thanks @eeberhard that's true that this now makes more sense. I'd appreciate if we put a comment there for future reference explaining all this.

**Review by bpapaspyros — APPROVED** [review]
